In [1]:
import pandas as pd
import glob
import os

## 대민 데이터 결측치 / 이상치

In [2]:
import os
import pandas as pd

file_path = "../Data/신용정보.csv"

print(os.path.exists(file_path))

df_credit = pd.read_csv(
    file_path,
    encoding="utf-8-sig"
)

df_credit.head()

True


,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명
0,202301,41131,4113110100,20,2984,2874,66,신흥1동,성남시 수정구
1,202301,41131,4113110100,20,2984,2874,66,신흥2동,성남시 수정구
2,202301,41131,4113110100,20,2984,2874,66,신흥3동,성남시 수정구
3,202301,41131,4113110100,30,3523,3375,105,신흥1동,성남시 수정구
4,202301,41131,4113110100,30,3523,3375,105,신흥2동,성남시 수정구


In [3]:
df_credit.shape

(29818, 9)

In [7]:
# 결측치 보기
df_credit.isnull().sum()

BS_YR_MON    0
SIGUNGU      0
BCD          0
AGE          0
TOT_CNT      0
ECON_CNT     0
NECON_CNT    0
읍면동명         0
시군구명         0
dtype: int64

In [5]:
# 성남시가 아닌 경우
df_credit[~df_credit['시군구명'].str.contains('성남시', na=False)]

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명


In [6]:
df_credit['AGE'].value_counts()

AGE
20    4979
40    4971
50    4968
60    4968
70    4968
30    4964
Name: count, dtype: int64

In [7]:
df_credit['TOT_CNT'].describe()

count    29818.000000
mean      2195.166778
std       1694.543506
min          0.000000
25%        682.000000
50%       2090.000000
75%       3360.000000
max       7522.000000
Name: TOT_CNT, dtype: float64

In [8]:
df_credit

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명
0,202301,41131,4113110100,20,2984,2874,66,신흥1동,성남시 수정구
1,202301,41131,4113110100,20,2984,2874,66,신흥2동,성남시 수정구
2,202301,41131,4113110100,20,2984,2874,66,신흥3동,성남시 수정구
3,202301,41131,4113110100,30,3523,3375,105,신흥1동,성남시 수정구
4,202301,41131,4113110100,30,3523,3375,105,신흥2동,성남시 수정구
...,...,...,...,...,...,...,...,...,...
29813,202512,41135,4113511800,30,13,11,0,운중동,성남시 분당구
29814,202512,41135,4113511800,40,0,0,0,운중동,성남시 분당구
29815,202512,41135,4113511800,50,0,0,0,운중동,성남시 분당구
29816,202512,41135,4113511800,60,0,0,0,운중동,성남시 분당구


29818 행 중 504행이 총인구수, 경제활동참여자수, 비경제활동참여수가 0이다.
     - 성남시 분당구 운중동이 486행 나머지는 성남시 수정구 고등동에 분포함

In [8]:
# 읍면동 = 행정동 
df_credit[df_credit['TOT_CNT'] == 0]

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명
810,202301,41135,4113511700,20,0,0,0,운중동,성남시 분당구
816,202301,41135,4113511800,20,0,0,0,운중동,성남시 분당구
817,202301,41135,4113511800,40,0,0,0,운중동,성남시 분당구
818,202301,41135,4113511800,50,0,0,0,운중동,성남시 분당구
819,202301,41135,4113511800,60,0,0,0,운중동,성남시 분당구
...,...,...,...,...,...,...,...,...,...
29812,202512,41135,4113511800,20,0,0,0,운중동,성남시 분당구
29814,202512,41135,4113511800,40,0,0,0,운중동,성남시 분당구
29815,202512,41135,4113511800,50,0,0,0,운중동,성남시 분당구
29816,202512,41135,4113511800,60,0,0,0,운중동,성남시 분당구


In [9]:
df_credit[
    (df_credit['시군구명'] == '성남시 분당구') &
    (df_credit['읍면동명'] == '운중동') &
    (df_credit['TOT_CNT'] != 0)
]

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명
780,202301,41135,4113511500,20,1294,1181,74,운중동,성남시 분당구
781,202301,41135,4113511500,30,1306,1238,55,운중동,성남시 분당구
782,202301,41135,4113511500,40,1809,1770,22,운중동,성남시 분당구
783,202301,41135,4113511500,50,1731,1680,35,운중동,성남시 분당구
784,202301,41135,4113511500,60,1219,1171,34,운중동,성남시 분당구
...,...,...,...,...,...,...,...,...,...
29799,202512,41135,4113511700,70,15,12,3,운중동,성남시 분당구
29803,202512,41135,4113511700,50,15,14,0,운중동,성남시 분당구
29804,202512,41135,4113511700,60,15,14,0,운중동,성남시 분당구
29805,202512,41135,4113511700,70,19,13,4,운중동,성남시 분당구


In [11]:
df_credit[
    (df_credit['TOT_CNT'] == 0) &
    (df_credit['ECON_CNT'] == 0) &
    (df_credit['NECON_CNT'] == 0)
][['BS_YR_MON', 'BCD', 'AGE', '읍면동명', '시군구명']]

,BS_YR_MON,BCD,AGE,읍면동명,시군구명
810,202301,4113511700,20,운중동,성남시 분당구
816,202301,4113511800,20,운중동,성남시 분당구
817,202301,4113511800,40,운중동,성남시 분당구
818,202301,4113511800,50,운중동,성남시 분당구
819,202301,4113511800,60,운중동,성남시 분당구
...,...,...,...,...,...
29812,202512,4113511800,20,운중동,성남시 분당구
29814,202512,4113511800,40,운중동,성남시 분당구
29815,202512,4113511800,50,운중동,성남시 분당구
29816,202512,4113511800,60,운중동,성남시 분당구


In [10]:
df42 = df_credit[
    (df_credit['읍면동명'] == '운중동') &
    (df_credit['TOT_CNT'] == 0) &
    (df_credit['ECON_CNT'] == 0) &
    (df_credit['NECON_CNT'] == 0)
]
df42

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명
810,202301,41135,4113511700,20,0,0,0,운중동,성남시 분당구
816,202301,41135,4113511800,20,0,0,0,운중동,성남시 분당구
817,202301,41135,4113511800,40,0,0,0,운중동,성남시 분당구
818,202301,41135,4113511800,50,0,0,0,운중동,성남시 분당구
819,202301,41135,4113511800,60,0,0,0,운중동,성남시 분당구
...,...,...,...,...,...,...,...,...,...
29812,202512,41135,4113511800,20,0,0,0,운중동,성남시 분당구
29814,202512,41135,4113511800,40,0,0,0,운중동,성남시 분당구
29815,202512,41135,4113511800,50,0,0,0,운중동,성남시 분당구
29816,202512,41135,4113511800,60,0,0,0,운중동,성남시 분당구


In [13]:
df_credit[df_credit['BCD'] == 4113511700]

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명
804,202301,41135,4113511700,20,18,17,0,운중동,성남시 분당구
805,202301,41135,4113511700,30,17,17,0,운중동,성남시 분당구
806,202301,41135,4113511700,40,15,15,0,운중동,성남시 분당구
807,202301,41135,4113511700,50,17,16,0,운중동,성남시 분당구
808,202301,41135,4113511700,60,19,19,0,운중동,성남시 분당구
...,...,...,...,...,...,...,...,...,...
29801,202512,41135,4113511700,30,0,0,0,운중동,성남시 분당구
29802,202512,41135,4113511700,40,0,0,0,운중동,성남시 분당구
29803,202512,41135,4113511700,50,15,14,0,운중동,성남시 분당구
29804,202512,41135,4113511700,60,15,14,0,운중동,성남시 분당구


In [15]:
df42['AGE'].value_counts()

AGE
20    103
40     92
50     78
70     78
60     72
30     63
Name: count, dtype: int64

성남시 분당구 운중동이 486행 나머지는 성남시 수정구 고등동에 분포함


In [16]:
df_credit[
    (df_credit['읍면동명'] != '운중동') &
    (df_credit['TOT_CNT'] == 0) &
    (df_credit['ECON_CNT'] == 0) &
    (df_credit['NECON_CNT'] == 0)
]

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명
6028,202308,41131,4113111400,20,0,0,0,고등동,성남시 수정구
6857,202309,41131,4113111400,20,0,0,0,고등동,성남시 수정구
7686,202310,41131,4113111400,20,0,0,0,고등동,성남시 수정구
8515,202311,41131,4113111400,20,0,0,0,고등동,성남시 수정구
9344,202312,41131,4113111400,20,0,0,0,고등동,성남시 수정구
10173,202401,41131,4113111400,20,0,0,0,고등동,성남시 수정구
11002,202402,41131,4113111400,20,0,0,0,고등동,성남시 수정구
11831,202403,41131,4113111400,20,0,0,0,고등동,성남시 수정구
12660,202404,41131,4113111400,20,0,0,0,고등동,성남시 수정구
13489,202405,41131,4113111400,20,0,0,0,고등동,성남시 수정구


### 504행이 총인구수, 경제활동참여자수, 비경제활동참여수가 0인 데이터 처리 전략

- 확인 결과: 전체 29,818행 중 504행에서 `TOT_CNT`, `ECON_CNT`, `NECON_CNT`가 모두 0으로 나타남

- 위치 특성: 앞선 확인 결과 기준으로 대부분은 성남시 분당구 운중동, 나머지는 성남시 수정구 고등동에 분포함



#### 1. 이 값이 의미하는 것

- `TOT_CNT` = 총인구수

- `ECON_CNT` = 경제활동참여자수

- `NECON_CNT` = 비경제활동참여자수

- 세 값이 모두 0이라는 것은 해당 시점·행정동·연령대 조합에서 **집계된 인구가 없다고 기록되었다**는 뜻임



#### 2. 집계 정정성 판단

초보자 기준에서는 아래처럼 해석하면 됨.



- **바로 오류라고 단정하면 안 됨**

  - 실제로 해당 조건의 인구가 없어서 0일 수도 있음

- **하지만 이상치 후보로는 봐야 함**

  - 일반적으로 인구 관련 데이터에서 특정 지역에 0값이 반복적으로 몰려 있으면

  - 집계 누락, 입력 오류, 수집 기준 변경 가능성도 함께 의심해야 함

- 특히 이번 데이터는 0값이 특정 동에 집중되어 있어

  - 정상적인 0인지

  - 아니면 특정 지역 데이터가 비어 있는 것인지

  - 추가 확인이 필요함



#### 3. 처리 방향

1. **원본 자료와 먼저 대조하기**

   - 가장 좋은 방법은 원본 파일, 공공데이터 원자료, 같은 기간의 다른 인구 자료와 비교하는 것임

   - 원본에서도 0이면 실제 0일 가능성이 큼

   - 원본에는 값이 있는데 현재 파일만 0이면 전처리 또는 입력 문제일 가능성이 큼



2. **임의로 평균값으로 채우지 않기**

   - 인구 데이터는 지역·월·연령대 특성이 커서

   - 근거 없이 평균이나 주변 값으로 대체하면 왜곡될 수 있음



3. **분석 목적에 따라 구분 처리하기**

   - 기초 탐색(EDA) 단계: `이상치 후보` 또는 `확인 필요` 라벨을 붙여 따로 관리

   - 모델링 단계: 해당 행을 제외할지, 별도 플래그 변수를 만들지 결정

   - 보고서 단계: `특정 지역에 0값이 집중되어 해석에 주의가 필요함`이라고 명시



4. **세 값이 모두 0인 경우만 묶어서 관리하기**

   - 이번처럼 `TOT_CNT == 0`, `ECON_CNT == 0`, `NECON_CNT == 0`을 동시에 만족하는 행은

   - 단순 결측치보다 `구조적 0 또는 집계 이상` 후보로 따로 보는 것이 좋음



#### 4. 초보자용 결론

- 이 504행은 **무조건 삭제할 데이터**도 아니고, **무조건 정상 데이터**도 아님

- 가장 안전한 방향은

  - 먼저 원본과 대조하고

  - 원인 확인 전까지는 `이상치 후보`로 표시해 두는 것임

- 즉, 현재 단계의 권장 처리 방향은 **성급히 값 대체하지 말고, 별도 표시 후 검증**임


In [4]:
import numpy as np
key_cols = ['BS_YR_MON', 'BCD', 'AGE', '읍면동명', '시군구명']

# 1. key별 0값 상태 집계
grp = (
    df_credit.groupby(key_cols, dropna=False)
      .agg(
          row_cnt=('TOT_CNT', 'size'),
          all_zero=('TOT_CNT', lambda x: (x == 0).all()),
          any_zero=('TOT_CNT', lambda x: (x == 0).any()),
          any_nonzero=('TOT_CNT', lambda x: (x > 0).any())
      )
      .reset_index()
)

# 2. zero 유형 분류
grp['zero_type'] = np.select(
    [
        grp['any_zero'] & grp['any_nonzero'],
        grp['all_zero']
    ],
    [
        'error_zero',
        'structural_zero'
    ],
    default='normal'
)

# 3. 원본 데이터에 zero 유형 붙이기
df2 = df_credit.merge(
    grp[key_cols + ['zero_type']],
    on=key_cols,
    how='left'
)

# 4. 분석용 clean 데이터 생성
# - error_zero: 제거
# - structural_zero: 분석용에서 제외
clean_df = df2[
    ~(
        (df2['TOT_CNT'] == 0) &
        (df2['zero_type'].isin(['error_zero', 'structural_zero']))
    )
].copy()

# 5. 감사용 로그
zero_log = df2[df2['TOT_CNT'] == 0].copy()

### data\경기도_주민등록인구통계성연령별인구현황행정동별집계기본.csv 파일로 해당 데이터가 실제 0인지 점검 가능한가?

가능하다. 다만 **바로 1:1 값 비교를 하는 방식보다는, 행정동·기준월·연령대 기준으로 비교 가능한지부터 확인한 뒤 대조**하는 것이 안전하다.

#### 비교 가능 여부를 먼저 확인할 항목

- **공통 키 존재 여부 확인**
  - 현재 신용 데이터의 비교 키는 `BS_YR_MON`, `BCD`, `AGE`, `읍면동명`, `시군구명`이다.
  - 주민등록인구 파일에도 이와 대응되는 `기준월`, `행정동코드`, `연령`, `행정동명`, `시군구명` 성격의 컬럼이 있는지 먼저 확인해야 한다.

- **행정동 기준 일치 여부 확인**
  - 두 파일이 모두 행정동 기준인지 확인해야 한다.
  - 한쪽은 법정동, 다른 한쪽은 행정동이면 바로 비교하면 안 된다.

- **시점 기준 일치 여부 확인**
  - `BS_YR_MON`과 주민등록 파일의 월 기준이 같은 형식인지 확인한다.
  - 예: `202401`, `2024-01`, `2024년 1월`처럼 표현 방식이 다를 수 있으므로 통일이 필요하다.

- **연령 구간 기준 일치 여부 확인**
  - 신용 데이터의 `AGE`가 단일 나이인지, 연령대(예: 20대, 30대)인지 확인해야 한다.
  - 주민등록 파일이 `만 나이`, `5세 단위`, `10세 단위` 집계이면 재분류 후 비교해야 한다.

#### 비교 전략

1. **비교용 공통 키를 만든다**

   - 두 데이터에서 아래 기준으로 비교 키를 맞춘다.
   - `기준월 + 행정동코드(또는 행정동명) + 연령 기준`

2. **신용 데이터의 0건 행만 따로 추출한다**

   - 특히 `TOT_CNT == 0`, `ECON_CNT == 0`, `NECON_CNT == 0`인 행을 우선 점검 대상으로 둔다.
   - 운중동, 고등동처럼 0이 몰린 동을 우선 확인하면 된다.

3. **주민등록 파일과 매칭한다**

   - 같은 월·같은 행정동·같은 연령 기준으로 주민등록 인구 수를 붙인다.
   - 이때 매칭 성공률이 낮으면 키 정의부터 다시 점검해야 한다.

4. **비교 결과를 3가지로 나눈다**

   - **주민등록도 0 또는 매우 작음**
     - 실제 인구가 거의 없는 구간일 수 있으므로 `구조적 0` 가능성이 있다.

   - **주민등록은 0이 아닌데 신용 데이터만 0**
     - 집계 누락, 전처리 오류, 매핑 오류 가능성이 크므로 `이상치 후보`로 보는 것이 타당하다.

   - **아예 매칭되지 않음**
     - 코드 체계 불일치, 행정동명 표기 차이, 월 형식 차이 가능성이 있으므로 정합성 문제를 먼저 해결해야 한다.

#### 해석 시 주의점

- 주민등록인구는 **전체 등록인구**이고, 신용 데이터의 `TOT_CNT`는 **해당 데이터 공급기관 기준 집계 인구**일 수 있다.
- 따라서 두 값이 **정확히 같아야 하는 것은 아니고**, 여기서는 `0이 실제로 가능한가`를 점검하는 보조 근거로 활용하는 것이 적절하다.
- 즉, **절대값 일치 검증**보다 **0 대 비0 검증**에 더 유용하다.

#### 결론

- `data\경기도_주민등록인구통계성연령별인구현황행정동별집계기본.csv`로 **해당 0값이 실제 0인지 점검하는 것은 가능하다**.
- 다만 바로 단순 비교하기보다,
  - 행정동 기준 일치 여부,
  - 월 기준 통일,
  - 연령 구간 재정렬
  를 먼저 맞춘 뒤 비교해야 한다.
- 가장 실무적인 방법은 **신용 데이터의 0건 행을 추출한 뒤 주민등록 데이터와 매칭해서 `실제 0 가능`, `이상치 후보`, `매칭불가`로 분류하는 것**이다.

In [19]:
# 주민등록인구 파일을 이용해 신용 데이터의 0값이 '실제 0 가능성'이 있는지 1차 점검
# 주의: 이 파일은 연도/행정동 기준 비교에 더 적합하므로, 월·연령까지 완전 일치 검증은 아님

resident_path = r"C:\Users\cute3\OneDrive\Desktop\pro\final_project\Data\202301_202306_연령별인구현황_월간.csv"

resident_raw = None
used_encoding = None

for enc in ['cp949', 'euc-kr', 'utf-8-sig']:
    try:
        resident_raw = pd.read_csv(resident_path, encoding=enc)
        used_encoding = enc
        break
    except Exception:
        pass

if resident_raw is None:
    raise FileNotFoundError(f"주민등록 파일을 읽지 못했습니다: {resident_path}")

print(f"주민등록 파일 로드 완료 / encoding = {used_encoding}")

# 헤더 인코딩 상태와 무관하게 앞 3개 컬럼을 비교 키로 사용
resident = resident_raw.iloc[:, :3].copy()
resident.columns = ['resident_year', 'resident_bcd_raw', '읍면동명']

resident['resident_year'] = (
    resident['resident_year']
    .astype(str)
    .str.extract(r'(\d{4})', expand=False)
    .astype('Int64')
)
resident['resident_bcd_8'] = (
    resident['resident_bcd_raw']
    .astype(str)
    .str.extract(r'(\d+)', expand=False)
    .str[:8]
    .fillna('')
)
resident['읍면동명'] = resident['읍면동명'].astype(str).str.strip()

# 숫자형 인구 컬럼 전체를 사용해 '해당 동에 인구가 하나라도 집계되었는가'를 확인
resident_num = resident_raw.iloc[:, 3:].apply(pd.to_numeric, errors='coerce')
resident['resident_any_pop'] = resident_num.fillna(0).max(axis=1) > 0

# 연도 + 행정동 기준으로 요약
resident_check = (
    resident.groupby(['resident_year', '읍면동명'], dropna=False)
    .agg(
        resident_row_cnt=('resident_any_pop', 'size'),
        resident_any_pop=('resident_any_pop', 'max'),
        resident_bcd_candidates=('resident_bcd_8', lambda x: ', '.join(sorted(set(v for v in x if v))))
    )
    .reset_index()
    .rename(columns={'읍면동명': '읍면동명_resident'})
)

# 신용 데이터에서 triple zero 대상 추출
credit_zero_check = df_credit[
    (df_credit['TOT_CNT'] == 0) &
    (df_credit['ECON_CNT'] == 0) &
    (df_credit['NECON_CNT'] == 0)
].copy()

credit_zero_check['resident_year'] = (
    credit_zero_check['BS_YR_MON']
    .astype(str)
    .str.extract(r'(\d{4})', expand=False)
    .astype('Int64')
)
credit_zero_check['credit_bcd_8'] = (
    credit_zero_check['BCD']
    .astype(str)
    .str.extract(r'(\d+)', expand=False)
    .str[:8]
    .fillna('')
)
credit_zero_check['읍면동명'] = credit_zero_check['읍면동명'].astype(str).str.strip()

compare_zero = credit_zero_check.merge(
    resident_check,
    left_on=['resident_year', '읍면동명'],
    right_on=['resident_year', '읍면동명_resident'],
    how='left'
 )

compare_zero['resident_code_match'] = compare_zero.apply(
    lambda row: row['credit_bcd_8'] in str(row['resident_bcd_candidates']).split(', ')
    if pd.notna(row['resident_bcd_candidates']) and str(row['resident_bcd_candidates']).strip() != ''
    else False,
    axis=1
)

resident_any_pop_flag = compare_zero['resident_any_pop'].fillna(False).astype(bool)

compare_zero['resident_check_result'] = np.select(
    [
        compare_zero['resident_row_cnt'].isna(),
        resident_any_pop_flag,
        ~resident_any_pop_flag
    ],
    [
        '매칭불가',
        '주민등록상 인구 있음 → 신용 0 재점검 필요',
        '주민등록상도 0 또는 미집계 가능'
    ],
    default='확인 필요'
 )

summary = compare_zero['resident_check_result'].value_counts(dropna=False).rename_axis('판정').reset_index(name='건수')
print('\n[주민등록 파일 기준 1차 점검 결과]')
display(summary)

print('\n[행정동별 요약]')
display(
    compare_zero.groupby(['시군구명', '읍면동명', 'resident_check_result'], dropna=False)
    .size()
    .reset_index(name='건수')
    .sort_values(['건수', '시군구명', '읍면동명'], ascending=[False, True, True])
    .head(20)
 )

print('\n[상세 확인용 샘플]')
display(
    compare_zero[[
        'BS_YR_MON', '시군구명', '읍면동명', 'BCD', 'AGE',
        'TOT_CNT', 'ECON_CNT', 'NECON_CNT',
        'resident_year', 'credit_bcd_8', 'resident_bcd_candidates',
        'resident_code_match', 'resident_check_result'
    ]]
    .sort_values(['resident_check_result', '시군구명', '읍면동명', 'BS_YR_MON', 'AGE'])
    .head(50)
 )

compare_zero.head()

주민등록 파일 로드 완료 / encoding = cp949

[주민등록 파일 기준 1차 점검 결과]


,판정,건수
0,매칭불가,504



[행정동별 요약]


,시군구명,읍면동명,resident_check_result,건수
0,성남시 분당구,운중동,매칭불가,486
1,성남시 수정구,고등동,매칭불가,18



[상세 확인용 샘플]


,BS_YR_MON,시군구명,읍면동명,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,resident_year,credit_bcd_8,resident_bcd_candidates,resident_code_match,resident_check_result
0,202301,성남시 분당구,운중동,4113511700,20,0,0,0,2023,41135117,NaN,False,매칭불가
1,202301,성남시 분당구,운중동,4113511800,20,0,0,0,2023,41135118,NaN,False,매칭불가
6,202301,성남시 분당구,운중동,4113511800,30,0,0,0,2023,41135118,NaN,False,매칭불가
2,202301,성남시 분당구,운중동,4113511800,40,0,0,0,2023,41135118,NaN,False,매칭불가
7,202301,성남시 분당구,운중동,4113511800,40,0,0,0,2023,41135118,NaN,False,매칭불가
3,202301,성남시 분당구,운중동,4113511800,50,0,0,0,2023,41135118,NaN,False,매칭불가
8,202301,성남시 분당구,운중동,4113511800,50,0,0,0,2023,41135118,NaN,False,매칭불가
4,202301,성남시 분당구,운중동,4113511800,60,0,0,0,2023,41135118,NaN,False,매칭불가
9,202301,성남시 분당구,운중동,4113511800,60,0,0,0,2023,41135118,NaN,False,매칭불가
5,202301,성남시 분당구,운중동,4113511800,70,0,0,0,2023,41135118,NaN,False,매칭불가


,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명,resident_year,credit_bcd_8,읍면동명_resident,resident_row_cnt,resident_any_pop,resident_bcd_candidates,resident_code_match,resident_check_result
0,202301,41135,4113511700,20,0,0,0,운중동,성남시 분당구,2023,41135117,NaN,NaN,NaN,NaN,False,매칭불가
1,202301,41135,4113511800,20,0,0,0,운중동,성남시 분당구,2023,41135118,NaN,NaN,NaN,NaN,False,매칭불가
2,202301,41135,4113511800,40,0,0,0,운중동,성남시 분당구,2023,41135118,NaN,NaN,NaN,NaN,False,매칭불가
3,202301,41135,4113511800,50,0,0,0,운중동,성남시 분당구,2023,41135118,NaN,NaN,NaN,NaN,False,매칭불가
4,202301,41135,4113511800,60,0,0,0,운중동,성남시 분당구,2023,41135118,NaN,NaN,NaN,NaN,False,매칭불가


In [22]:
# 1차 점검 결과가 모두 '매칭불가'인 이유를 추가 진단
# 핵심: 주민등록 파일의 연도 범위와 신용 0데이터의 연도 범위가 실제로 겹치는지 확인

credit_zero_diag = df_credit[
    (df_credit['TOT_CNT'] == 0) &
    (df_credit['ECON_CNT'] == 0) &
    (df_credit['NECON_CNT'] == 0)
].copy()

credit_zero_diag['credit_year'] = (
    credit_zero_diag['BS_YR_MON']
    .astype(str)
    .str.extract(r'(\d{4})', expand=False)
    .astype('Int64')
)
credit_zero_diag['credit_bcd_8'] = (
    credit_zero_diag['BCD']
    .astype(str)
    .str.extract(r'(\d+)', expand=False)
    .str[:8]
    .fillna('')
)

resident_diag = resident.copy()
resident_diag['resident_sigungu_5'] = resident_diag['resident_bcd_8'].str[:5]
credit_zero_diag['credit_sigungu_5'] = credit_zero_diag['credit_bcd_8'].str[:5]

credit_years = sorted([int(x) for x in credit_zero_diag['credit_year'].dropna().unique().tolist()])
resident_years = sorted([int(x) for x in resident_diag['resident_year'].dropna().unique().tolist()])
year_overlap = sorted(set(credit_years).intersection(resident_years))

print('[연도 범위 비교]')
display(pd.DataFrame({
    '구분': ['신용 0인 데이터', '주민등록 파일'],
    '연도범위': [
        f"{min(credit_years)} ~ {max(credit_years)}" if credit_years else '없음',
        f"{min(resident_years)} ~ {max(resident_years)}" if resident_years else '없음'
    ],
    '연도목록': [str(credit_years), str(resident_years)]
}))

print('[연도 교집합]')
print(year_overlap if year_overlap else '없음 → 현재 파일로는 같은 시점 비교 불가')

print('\n[신용 0인 데이터 위치 요약]')
display(
    credit_zero_diag.groupby(['credit_year', '시군구명', '읍면동명', 'credit_bcd_8'], dropna=False)
    .size()
    .reset_index(name='건수')
    .sort_values(['credit_year', '건수', '시군구명', '읍면동명'], ascending=[True, False, True, True])
 )

print('\n[주민등록 파일 내 동일 시군구 prefix 존재 여부]')
display(
    pd.DataFrame({
        'credit_sigungu_5': sorted(credit_zero_diag['credit_sigungu_5'].dropna().unique()),
    }).assign(
        resident_prefix_exists=lambda x: x['credit_sigungu_5'].isin(resident_diag['resident_sigungu_5'].dropna().unique())
    )
 )

if not year_overlap:
    print('\n[판단] 주민등록 파일은 지역 코드 자체는 일부 겹칠 수 있지만, 연도 범위가 2016~2018로 한정되어 있어 2023~2025 신용 0인 데이터의 실제 0 여부를 직접 검증하는 근거로 쓰기 어렵습니다.')
    print('따라서 같은 형식의 최신 주민등록 원자료(2023~2025)가 필요합니다.')
else:
    print('\n[판단] 일부 연도는 비교 가능하므로 교집합 연도만 추려서 재매칭하면 됩니다.')

[연도 범위 비교]


,구분,연도범위,연도목록
0,신용 0인 데이터,2023 ~ 2025,"[2023, 2024, 2025]"
1,주민등록 파일,4113 ~ 4113,[4113]


[연도 교집합]
없음 → 현재 파일로는 같은 시점 비교 불가

[신용 0인 데이터 위치 요약]


,credit_year,시군구명,읍면동명,credit_bcd_8,건수
1,2023,성남시 분당구,운중동,41135118,131
0,2023,성남시 분당구,운중동,41135117,24
2,2023,성남시 수정구,고등동,41131114,5
4,2024,성남시 분당구,운중동,41135118,135
3,2024,성남시 분당구,운중동,41135117,28
5,2024,성남시 수정구,고등동,41131114,13
7,2025,성남시 분당구,운중동,41135118,133
6,2025,성남시 분당구,운중동,41135117,35



[주민등록 파일 내 동일 시군구 prefix 존재 여부]


,credit_sigungu_5,resident_prefix_exists
0,41131,False
1,41135,False



[판단] 주민등록 파일은 지역 코드 자체는 일부 겹칠 수 있지만, 연도 범위가 2016~2018로 한정되어 있어 2023~2025 신용 0인 데이터의 실제 0 여부를 직접 검증하는 근거로 쓰기 어렵습니다.
따라서 같은 형식의 최신 주민등록 원자료(2023~2025)가 필요합니다.


In [5]:
# 재업로드한 2023년 1~6월 월간 연령별 인구 파일로 다시 점검
# 이 파일은 2023-01 ~ 2023-06 구간의 '운중동' zero 데이터 검증에 직접 활용 가능

resident_month_path = r"C:\Users\cute3\OneDrive\Desktop\pro\final_project\Data\202301_202306_연령별인구현황_월간.csv"

resident_monthly_raw = None
resident_monthly_encoding = None

for enc in ['cp949', 'euc-kr', 'utf-8-sig']:
    try:
        resident_monthly_raw = pd.read_csv(resident_month_path, encoding=enc)
        resident_monthly_encoding = enc
        break
    except Exception:
        pass

if resident_monthly_raw is None:
    raise FileNotFoundError(f"월간 주민등록 파일을 읽지 못했습니다: {resident_month_path}")

print(f"월간 주민등록 파일 로드 완료 / encoding = {resident_monthly_encoding}")

resident_monthly = resident_monthly_raw.copy()
first_col = resident_monthly.columns[0]
resident_monthly = resident_monthly.rename(columns={first_col: '행정구역'})

resident_monthly['resident_code'] = resident_monthly['행정구역'].astype(str).str.extract(r'\((\d{10})\)', expand=False)
resident_monthly['읍면동명'] = resident_monthly['행정구역'].astype(str).str.extract(r'([가-힣]+)\(\d{10}\)$', expand=False)

target_age_map = {
    '18~29세': 20,
    '30~39세': 30,
    '40~49세': 40,
    '50~59세': 50,
    '60~69세': 60,
    '70~79세': 70
}

target_cols = [
    col for col in resident_monthly.columns
    if any(age_band in col for age_band in target_age_map.keys()) and ('_거주자_' in col) and ('남_' not in col) and ('여_' not in col)
 ]

resident_long = resident_monthly.melt(
    id_vars=['행정구역', 'resident_code', '읍면동명'],
    value_vars=target_cols,
    var_name='resident_col',
    value_name='resident_pop_raw'
 )

resident_long['월'] = resident_long['resident_col'].str.extract(r'2023년(\d{2})월', expand=False)
resident_long['BS_YR_MON'] = ('2023' + resident_long['월']).astype(int)
resident_long['age_band'] = resident_long['resident_col'].str.extract(r'(20~29세|30~39세|40~49세|50~59세|60~69세|70~79세)', expand=False)
resident_long['AGE'] = resident_long['age_band'].map(target_age_map)
resident_long['resident_pop'] = (
    resident_long['resident_pop_raw']
    .astype(str)
    .str.replace(',', '', regex=False)
    .astype(int)
 )

resident_long = resident_long[['행정구역', 'resident_code', '읍면동명', 'BS_YR_MON', 'AGE', 'resident_pop']]

credit_zero_h1 = df_credit[
    (df_credit['TOT_CNT'] == 0) &
    (df_credit['ECON_CNT'] == 0) &
    (df_credit['NECON_CNT'] == 0) &
    (df_credit['BS_YR_MON'].between(202301, 202306))
].copy()

compare_zero_h1 = credit_zero_h1.merge(
    resident_long,
    on=['읍면동명', 'BS_YR_MON', 'AGE'],
    how='left'
 )

compare_zero_h1['resident_check_result'] = np.select(
    [
        compare_zero_h1['resident_pop'].isna(),
        compare_zero_h1['resident_pop'] > 0,
        compare_zero_h1['resident_pop'] == 0
    ],
    [
        '매칭불가',
        '주민등록상 인구 있음 → 신용 0 재점검 필요',
        '주민등록상도 0 가능'
    ],
    default='확인 필요'
 )

print('\n[2023년 1~6월 기준 재점검 결과]')
display(
    compare_zero_h1['resident_check_result']
    .value_counts(dropna=False)
    .rename_axis('판정')
    .reset_index(name='건수')
 )

print('\n[동별 / 코드별 요약]')
display(
    compare_zero_h1.groupby(['시군구명', '읍면동명', 'BCD', 'resident_code', 'resident_check_result'], dropna=False)
    .size()
    .reset_index(name='건수')
    .sort_values(['건수', '시군구명', '읍면동명', 'BCD'], ascending=[False, True, True, True])
 )

print('\n[월·연령대별 비교 요약]')
display(
    compare_zero_h1.groupby(['BS_YR_MON', 'AGE', 'resident_check_result'], dropna=False)
    .agg(
        credit_zero_rows=('TOT_CNT', 'size'),
        resident_pop_min=('resident_pop', 'min'),
        resident_pop_max=('resident_pop', 'max')
    )
    .reset_index()
    .sort_values(['BS_YR_MON', 'AGE'])
 )

print('\n[상세 비교 샘플]')
display(
    compare_zero_h1[[
        'BS_YR_MON', '시군구명', '읍면동명', 'BCD', 'AGE',
        'TOT_CNT', 'ECON_CNT', 'NECON_CNT',
        'resident_code', 'resident_pop', 'resident_check_result'
    ]]
    .sort_values(['BS_YR_MON', 'BCD', 'AGE'])
    .head(100)
 )

compare_zero_h1.head()

월간 주민등록 파일 로드 완료 / encoding = cp949

[2023년 1~6월 기준 재점검 결과]


,판정,건수
0,주민등록상 인구 있음 → 신용 0 재점검 필요,66
1,매칭불가,14



[동별 / 코드별 요약]


,시군구명,읍면동명,BCD,resident_code,resident_check_result,건수
2,성남시 분당구,운중동,4113511800,4113568000,주민등록상 인구 있음 → 신용 0 재점검 필요,56
0,성남시 분당구,운중동,4113511700,4113568000,주민등록상 인구 있음 → 신용 0 재점검 필요,10
3,성남시 분당구,운중동,4113511800,NaN,매칭불가,8
1,성남시 분당구,운중동,4113511700,NaN,매칭불가,6



[월·연령대별 비교 요약]


,BS_YR_MON,AGE,resident_check_result,credit_zero_rows,resident_pop_min,resident_pop_max
0,202301,20,매칭불가,2,NaN,NaN
1,202301,30,주민등록상 인구 있음 → 신용 0 재점검 필요,1,5042.0,5042.0
2,202301,40,주민등록상 인구 있음 → 신용 0 재점검 필요,2,6336.0,6336.0
3,202301,50,주민등록상 인구 있음 → 신용 0 재점검 필요,2,5381.0,5381.0
4,202301,60,주민등록상 인구 있음 → 신용 0 재점검 필요,2,3642.0,3642.0
5,202301,70,주민등록상 인구 있음 → 신용 0 재점검 필요,2,1295.0,1295.0
6,202302,20,매칭불가,2,NaN,NaN
7,202302,30,주민등록상 인구 있음 → 신용 0 재점검 필요,1,5016.0,5016.0
8,202302,40,주민등록상 인구 있음 → 신용 0 재점검 필요,2,6369.0,6369.0
9,202302,50,주민등록상 인구 있음 → 신용 0 재점검 필요,3,5385.0,5385.0



[상세 비교 샘플]


,BS_YR_MON,시군구명,읍면동명,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,resident_code,resident_pop,resident_check_result
0,202301,성남시 분당구,운중동,4113511700,20,0,0,0,NaN,NaN,매칭불가
1,202301,성남시 분당구,운중동,4113511800,20,0,0,0,NaN,NaN,매칭불가
6,202301,성남시 분당구,운중동,4113511800,30,0,0,0,4113568000,5042.0,주민등록상 인구 있음 → 신용 0 재점검 필요
2,202301,성남시 분당구,운중동,4113511800,40,0,0,0,4113568000,6336.0,주민등록상 인구 있음 → 신용 0 재점검 필요
7,202301,성남시 분당구,운중동,4113511800,40,0,0,0,4113568000,6336.0,주민등록상 인구 있음 → 신용 0 재점검 필요
...,...,...,...,...,...,...,...,...,...,...,...
77,202306,성남시 분당구,운중동,4113511800,50,0,0,0,4113568000,5449.0,주민등록상 인구 있음 → 신용 0 재점검 필요
72,202306,성남시 분당구,운중동,4113511800,60,0,0,0,4113568000,3741.0,주민등록상 인구 있음 → 신용 0 재점검 필요
78,202306,성남시 분당구,운중동,4113511800,60,0,0,0,4113568000,3741.0,주민등록상 인구 있음 → 신용 0 재점검 필요
73,202306,성남시 분당구,운중동,4113511800,70,0,0,0,4113568000,1349.0,주민등록상 인구 있음 → 신용 0 재점검 필요


,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명,행정구역,resident_code,resident_pop,resident_check_result
0,202301,41135,4113511700,20,0,0,0,운중동,성남시 분당구,NaN,NaN,NaN,매칭불가
1,202301,41135,4113511800,20,0,0,0,운중동,성남시 분당구,NaN,NaN,NaN,매칭불가
2,202301,41135,4113511800,40,0,0,0,운중동,성남시 분당구,경기도 성남시 분당구 운중동(4113568000),4113568000,6336.0,주민등록상 인구 있음 → 신용 0 재점검 필요
3,202301,41135,4113511800,50,0,0,0,운중동,성남시 분당구,경기도 성남시 분당구 운중동(4113568000),4113568000,5381.0,주민등록상 인구 있음 → 신용 0 재점검 필요
4,202301,41135,4113511800,60,0,0,0,운중동,성남시 분당구,경기도 성남시 분당구 운중동(4113568000),4113568000,3642.0,주민등록상 인구 있음 → 신용 0 재점검 필요


In [6]:
clean_df.head()

,BS_YR_MON,SIGUNGU,BCD,AGE,TOT_CNT,ECON_CNT,NECON_CNT,읍면동명,시군구명,zero_type
0,202301,41131,4113110100,20,2984,2874,66,신흥1동,성남시 수정구,normal
1,202301,41131,4113110100,20,2984,2874,66,신흥2동,성남시 수정구,normal
2,202301,41131,4113110100,20,2984,2874,66,신흥3동,성남시 수정구,normal
3,202301,41131,4113110100,30,3523,3375,105,신흥1동,성남시 수정구,normal
4,202301,41131,4113110100,30,3523,3375,105,신흥2동,성남시 수정구,normal


## 전입 통계 전처리 / 이상치

In [ ]:
import pandas as pd
import numpy as np

df = df_credit.copy()

key_cols = ['BS_YR_MON', 'BCD', 'AGE', '읍면동명', '시군구명']

# 0 여부 플래그
df['triple_zero'] = (
    (df['TOT_CNT'] == 0) &
    (df['ECON_CNT'] == 0) &
    (df['NECON_CNT'] == 0)
)

# key별 상태 요약
grp = df.groupby(key_cols).agg(
    row_cnt=('TOT_CNT', 'size'),
    all_triple_zero=('triple_zero', 'all'),
    any_triple_zero=('triple_zero', 'any'),
    any_nonzero_tot=('TOT_CNT', lambda x: (x > 0).any())
).reset_index()

def classify_zero(row):
    if row['any_triple_zero'] and row['any_nonzero_tot']:
        return 'error_zero'
    elif row['all_triple_zero']:
        return 'structural_zero'
    else:
        return 'normal'

grp['zero_type'] = grp.apply(classify_zero, axis=1)

# 원본에 병합
df2 = df.merge(grp[key_cols + ['zero_type']], on=key_cols, how='left')

# 삭제 대신 우선 분석 제외 플래그
df2['exclude_from_analysis'] = (
    (df2['triple_zero']) &
    (df2['zero_type'].isin(['error_zero', 'structural_zero']))
)

# 분석용
clean_df = df2[~df2['exclude_from_analysis']].copy()

# 감사용 로그
zero_log = df2[df2['triple_zero']].copy()

In [ ]:
clean_df.head()

In [2]:
file_path = r"C:\Users\cute3\OneDrive\Desktop\pro\final_project\Data\전입통계.csv"

# 파일 존재 여부 먼저 확인
print(os.path.exists(file_path))

# True가 나오면 실행
df_in = pd.read_csv(file_path, encoding="utf-8-sig")

df_in.head(2)

True


,BF_YR_MON,BF_SIDO,BF_SIGNGU,BS_YR_MON,BS_SIGNGU,BS_HCD,TOT_CNT,HIGH_CNT,AVG_INC,AVG_INC_1,...,CNT_3,CNT_4,CNT_5,CNT_6,CNT_7,CNT_8,CNT_9,CNT_10,읍면동명,시군구명
0,202201,41,41131,202301,41111,4111156000,0,0,0,0,...,0,0,0,0,0,0,0,0,파장동,수원시 장안구
1,202201,41,41131,202301,41111,4111156000,0,0,0,0,...,0,0,0,0,0,0,0,0,파장동,수원시 장안구


In [4]:
df_in.shape

(405746, 31)

In [5]:
df_in.isnull().sum()

BF_YR_MON     0
BF_SIDO       0
BF_SIGNGU     0
BS_YR_MON     0
BS_SIGNGU     0
BS_HCD        0
TOT_CNT       0
HIGH_CNT      0
AVG_INC       0
AVG_INC_1     0
AVG_INC_2     0
AVG_INC_3     0
AVG_INC_4     0
AVG_INC_5     0
AVG_INC_6     0
AVG_INC_7     0
AVG_INC_8     0
AVG_INC_9     0
AVG_INC_10    0
CNT_1         0
CNT_2         0
CNT_3         0
CNT_4         0
CNT_5         0
CNT_6         0
CNT_7         0
CNT_8         0
CNT_9         0
CNT_10        0
읍면동명          0
시군구명          0
dtype: int64

## 전출 통계 이상치 / 결측치

In [6]:
file_path = r"C:\Users\cute3\OneDrive\Desktop\pro\final_project\Data\전출통계.csv"

# 파일 존재 여부 먼저 확인
print(os.path.exists(file_path))

# True가 나오면 실행
df_out = pd.read_csv(file_path, encoding="utf-8-sig")

df_out.head(2)

True


,BS_YR_MON,BS_SIGNGU,BS_HCD,AF_YR_MON,AF_SIDO,AF_SIGNGU,TOT_CNT,HIGH_CNT,AVG_INC,AVG_INC_1,...,CNT_3,CNT_4,CNT_5,CNT_6,CNT_7,CNT_8,CNT_9,CNT_10,읍면동명,시군구명
0,202201,41111,4111156000,202301,41,41131,0,0,0,0,...,0,0,0,0,0,0,0,0,파장동,수원시 장안구
1,202201,41111,4111156000,202301,41,41131,0,0,0,0,...,0,0,0,0,0,0,0,0,파장동,수원시 장안구


In [7]:
df_out.shape

(402335, 31)

In [8]:
df_out.isnull().sum()

BS_YR_MON     0
BS_SIGNGU     0
BS_HCD        0
AF_YR_MON     0
AF_SIDO       0
AF_SIGNGU     0
TOT_CNT       0
HIGH_CNT      0
AVG_INC       0
AVG_INC_1     0
AVG_INC_2     0
AVG_INC_3     0
AVG_INC_4     0
AVG_INC_5     0
AVG_INC_6     0
AVG_INC_7     0
AVG_INC_8     0
AVG_INC_9     0
AVG_INC_10    0
CNT_1         0
CNT_2         0
CNT_3         0
CNT_4         0
CNT_5         0
CNT_6         0
CNT_7         0
CNT_8         0
CNT_9         0
CNT_10        0
읍면동명          0
시군구명          0
dtype: int64